src/atmega2560p/hal/power.rs

```diff
@@ -0,0 +1,250 @@
+//     RustDuino : A generic HAL implementation for Arduino Boards in Rust
+//     Copyright (C) 2021  Devansh Kumar Jha,Indian Institute of Technology Kanpur
+//
+//     This program is free software: you can redistribute it and/or modify
+//     it under the terms of the GNU Affero General Public License as published
+//     by the Free Software Foundation, either version 3 of the License, or
+//     (at your option) any later version.
+//
+//     This program is distributed in the hope that it will be useful,
+//     but WITHOUT ANY WARRANTY; without even the implied warranty of
+//     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//     GNU Affero General Public License for more details.
+//
+//     You should have received a copy of the GNU Affero General Public License
+//     along with this program.  If not, see <https://www.gnu.org/licenses/>
+
+//! Generic implementation of power control through clock gating in ATMEGA2560P.
+//! Section 11.10.2 and 11.10.3 of the manual.
+//! Also references from Section 11.8.
+//! https://ww1.microchip.com/downloads/en/devicedoc/atmel-2549-8-bit-avr-microcontroller-atmega640-1280-1281-2560-2561_datasheet.pdf
+
+/// Crates required in the code for reading and writing to registers.
+use core;
+
+/// The options correspond to real world as shown -
+///  ```TWI    :  Power Reduction TWI```     
+///  ```TIMER2 :  Power Reduction Timer/Counter2```
+///  ```TIMER0 :  Power Reduction Timer/Counter0```
+///  ```TIMER1 :  Power Reduction Timer/Counter1```
+///  ```SPI    :  Power Reduction Serial Peripheral Interface```
+///  ```USART0 :  Power Reduction USART0```
+///  ```ADC    :  Power Reduction ADC```
+///  ```TIMER5 :  Power Reduction Timer/Counter5```
+///  ```TIMER4 :  Power Reduction Timer/Counter4```
+///  ```TIMER3 :  Power Reduction Timer/Counter3```
+///  ```USART3 :  Power Reduction USART3```
+///  ```USART2 :  Power Reduction USART2```  
+///  ```USART1 :  Power Reduction USART1```
+#[derive(Clone, Copy)]
+pub enum Options {
+    TWI,
+    TIMER2,
+    TIMER0,
+    TIMER1,
+    SPI,
+    USART0,
+    ADC,
+    TIMER5,
+    TIMER4,
+    TIMER3,
+    USART3,
+    USART2,
+    USART1,
+}
+
+/// Contains registers to control the functioning of clocks in the chip -
+/// PRR0 : Contains 7 Writable bits which will be used to control some of the clocks modes.
```

Comment:
I'm seeing register comments everywhere.

Reply:
Adjusted, sorry for bothering.